[MONDRIAN-2398] - Wrong results when aggregate tables used to calcula…

stanislau-strelchanka opened the pull request:
[MONDRIAN-2398] - Wrong results when aggregate tables used to calculate avg value which rolls up one or more NULLs

stanislau-strelchanka commented:
@kurtwalker , @lucboudreau , @mkambol , @mchen-len-son please review

wingman-pentaho commented:
### Build Skipped

:white_check_mark: There was nothing to build. The changes are probably excluded by the build rules. If this is an error, please contact your system administrator.

mkambol commented on src/main/mondrian/rolap/aggmatcher/ExplicitRecognizer.java:
```diff
@@ -440,6 +447,17 @@ public int compare(
+        String measureName = aggUsage.getColumn().getName();
+        Map<String, String> measuresFactCount = tableDef.getMeasuresFactCount();
+        String factCountColumnName = measuresFactCount.get(measureName);
+        if (StringUtils.isEmpty(factCountColumnName)) {
```
mondrian.olap.Util.isEmpty() is used in most places, probably best to stay consistent.

stanislau-strelchanka replied:
@mkambol replaced

mkambol commented on testsrc/main/mondrian/rolap/aggmatcher/AggMeasureFactCountTest.java:
```diff
@@ -0,0 +1,204 @@
+
+        verifySameAggAndNot(QUERY, schema);
+    }
+
```
I might be missing something, but these tests look identical to each other.
Also, they don't actually validate that an aggtable is used for the avg calculation.  You could verify that expected SQL is fired, instead.

stanislau-strelchanka replied:
@mkambol fixed tests and added generated sql validation

mkambol replied:
The following tests could still pass even if an agg table wasn't used.

mkambol commented on testsrc/main/mondrian/rolap/aggmatcher/AggMeasureFactCountTest.java:
```diff
@@ -0,0 +1,204 @@
+        propSaver.set(propSaver.properties.GenerateFormattedSql, true);
+        propSaver.set(propSaver.properties.UseAggregates, true);
+        propSaver.set(propSaver.properties.ReadAggregates, true);
+        propSaver.set(propSaver.properties.IgnoreInvalidMembers, true);
```
Why IgnoreInvalidMembers=true?

stanislau-strelchanka replied:
@mkambol removed this line

stanislau-strelchanka commented on src/main/mondrian/olap/Mondrian.xml:
```diff
@@ -1844,6 +1854,19 @@ Revision is $Id$
     <Element type="AggFactCount" class= "AggColumnName"/>
 
+    <Element type="AggMeasureFactCount" class= "AggColumnName">
+        <Attribute name="measure" required="false">
```
@mkambol renamed attribute to 'factColumn'

stanislau-strelchanka commented:
@mkambol I've added doc description for new element.
Regarding not using agg table if measure fact column is absent, I believe there might be cases 
where only one general fact count column will be enough for all fact table measures and there will be no need in creating multiple measure fact columns. Please correct me if I'm wrong.

stanislau-strelchanka commented:
@kurtwalker , @lucboudreau , @mkambol , @mchen-len-son please review

wingman-pentaho commented:
### Build Skipped

:white_check_mark: There was nothing to build. The changes are probably excluded by the build rules. If this is an error, please contact your system administrator.

mkambol commented on doc/schema.html:
```diff
@@ -3488,6 +3489,10 @@
+	<tr>
+		<td><code>&lt;<a name="XML_AggMeasureFactCount" href="xml_schema.html#AggMeasureFactCount">AggMeasureFactCount</a>&gt;</code></td>
+		<td>Specifies name of the column in the candidate aggregate table which contains the number of fact table rows associated with a specific measure.</td>
+	</tr>
```
To clarify intent, maybe something like:  "the number of non-null fact table rows associated with a specific fact column.  Mondrian can use this information when calculating averages from aggregate table values."

lucboudreau commented on testsrc/main/mondrian/rolap/aggmatcher/AggMeasureFactCountTest.java:
```diff
@@ -0,0 +1,288 @@
+                + "    <AggFactCount column=\"FACT_COUNT\"/>\n"
+                + "    <AggMeasureFactCount column=\"store_sales_fact_count\" factColumn=\"store_sales\" />\n"
+                + "    <AggMeasureFactCount column=\"store_cost_fact_count\" factColumn=\"store_cost\" />\n"
+                + "    <AggMeasureFactCount column=\"unit_sales_fact_count\" factColumn=\"unit_sales\" />\n"
```
Is the factColumn value supposed to be case sensitive? It is different here as compared to the definition within the AggMeasure element. What happens if the DB does not match those? Please try it / test it.

lucboudreau commented on src/main/mondrian/olap/Mondrian.xml:
```diff
@@ -1844,6 +1854,19 @@ Revision is $Id$
     <Element type="AggFactCount" class= "AggColumnName"/>
 
+    <Element type="AggMeasureFactCount" class= "AggColumnName">
+        <Attribute name="factColumn" required="false">
```
Why is factColumn not required? Are there scenarios where we don't set it? I also don't see any tests where it's not set.

wingman-pentaho commented:
### Build Skipped

:white_check_mark: There was nothing to build. The changes are probably excluded by the build rules. If this is an error, please contact your system administrator.

wingman-pentaho commented:
### Build Skipped

:white_check_mark: There was nothing to build. The changes are probably excluded by the build rules. If this is an error, please contact your system administrator.

wingman-pentaho commented:
### Build Completed

:x: This pull request has errors. They will need to be addressed before it can be accepted. See below for more details. Some links are also available below for further assistance in addressing those issues.

#### Build Commands

```sh
mvn -B -fn -DrunITs=true -f mondrian/pom.xml clean install && mvn -B -f mondrian/pom.xml site
mvn -B -fn -DrunITs=true -f workbench/pom.xml clean install && mvn -B -f workbench/pom.xml site
```

#### Cleanup Commands
```sh
mvn -B -f mondrian/pom.xml build-helper:remove-project-artifact && mvn -B -f workbench/pom.xml build-helper:remove-project-artifact
```

#### Changed files

```console
  mondrian/src/it/java/mondrian/rolap/aggmatcher/AggMeasureFactCountTest.java
  mondrian/src/it/java/mondrian/rolap/aggmatcher/agg_measure_fact_count_test.csv
  mondrian/src/main/java/mondrian/olap/Mondrian.xml
  mondrian/src/main/java/mondrian/rolap/aggmatcher/AggStar.java
  mondrian/src/main/java/mondrian/rolap/aggmatcher/DefaultRecognizer.java
  mondrian/src/main/java/mondrian/rolap/aggmatcher/ExplicitRecognizer.java
  mondrian/src/main/java/mondrian/rolap/aggmatcher/ExplicitRules.java
  mondrian/src/main/java/mondrian/rolap/aggmatcher/JdbcSchema.java
  mondrian/src/main/java/mondrian/rolap/aggmatcher/Recognizer.java
  mondrian/src/site/resources/doc/schema.html
  workbench/src/main/java/mondrian/gui/SchemaExplorer.java
  workbench/src/main/java/mondrian/gui/SchemaPropertyCellEditor.java
  workbench/src/main/java/mondrian/gui/SchemaTreeModel.java
  workbench/src/main/java/mondrian/olap/Mondrian_SW.xml
  workbench/src/main/resources/mondrian/gui/resources/text.properties
```

Still Broken Tests:
======================
###### mondrian.rolap.agg.SegmentBuilderTest.testSameRollupRegardlessOfSegmentOrderNoWildcards:

:red_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: 
SegmentMap is empty. No segmentIds matched test parameters. Full segment cache: [*Segment Header
Schema:[FoodMart]
Checksum:[7b4af973b0d21f364b0a746f5565cb03]
Cube:[Sales]
Measure:[Unit Sales]
Axes:[
    {product_class.product_department=('Alcoholic Beverages','Beverages','Periodicals')}
    {product_class.product_family=('Drink','Non-Consumable')}
    {time_by_day.the_year=('1997')}]
Excluded Regions:[]
Compound Predicates:[]
ID:[63b78b31fae7a3229f284be9a2
```
###### mondrian.rolap.agg.SegmentBuilderTest.testSameRollupRegardlessOfSegmentOrderThreeSegs:

:red_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: 
SegmentMap is empty. No segmentIds matched test parameters. Full segment cache: [*Segment Header
Schema:[FoodMart]
Checksum:[7b4af973b0d21f364b0a746f5565cb03]
Cube:[Sales]
Measure:[Unit Sales]
Axes:[
    {product_class.product_department=('Alcoholic Beverages','Periodicals')}
    {product_class.product_family=('Drink','Non-Consumable')}
    {time_by_day.the_year=('1997')}]
Excluded Regions:[]
Compound Predicates:[]
ID:[e546620837ba618128e2ffbb53f2a32867725f
```
###### mondrian.rolap.agg.SegmentBuilderTest.testSameRollupRegardlessOfSegmentOrderWithData:

:red_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: 
SegmentMap is empty. No segmentIds matched test parameters. Full segment cache: [*Segment Header
Schema:[FoodMart]
Checksum:[7b4af973b0d21f364b0a746f5565cb03]
Cube:[Sales]
Measure:[Unit Sales]
Axes:[
    {product_class.product_department=('Alcoholic Beverages','Baked Goods','Beverages','Periodicals')}
    {product_class.product_family=(*)}
    {time_by_day.the_year=('1997')}]
Excluded Regions:[]
Compound Predicates:[]
ID:[71cf49a1d6f0c2da433442927e05ea5c571
```
###### mondrian.rolap.agg.SegmentBuilderTest.testSameRollupRegardlessOfSegmentOrderWithEmptySegmentBody:

:red_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: 
SegmentMap is empty. No segmentIds matched test parameters. Full segment cache: [*Segment Header
Schema:[FoodMart]
Checksum:[7b4af973b0d21f364b0a746f5565cb03]
Cube:[Sales]
Measure:[Unit Sales]
Axes:[
    {time_by_day.month_of_year=('2','3','4','5','6','7')}
    {time_by_day.quarter=('Q1','Q2','Q3')}
    {time_by_day.the_year=('1998')}]
Excluded Regions:[]
Compound Predicates:[]
ID:[af45d690c892633e5755b06d623ba50a849357832d2157954303a76b4d9564ac]
, *Segment
```

#### Integration test coverage change

These statistics help you identify how your changes have affected the coverage of the following files. If a file is not in this list, then its coverage was not affected by your changes. To get some help interpreting these metrics, please refer to [Jacoco's documentation](http://eclemma.org/jacoco/trunk/doc/counters.html).

###### mondrian.olap.MondrianDef
 * Instruction Change:  + .02%

###### mondrian.olap.MondrianDef.AggName
 * Branch Change: **-1.32%**:small_red_triangle_down:
 * Complexity Change: **-3.24%**:small_red_triangle_down:
 * Instruction Change: **-.96%**:small_red_triangle_down:
 * Line Change:  + .02%

###### mondrian.olap.MondrianDef.AggTable
 * Complexity Change: **-.26%**:small_red_triangle_down:
 * Instruction Change: **-.13%**:small_red_triangle_down:
 * Line Change: **-.16%**:small_red_triangle_down:
 * Method Change:  + 3.21%

###### mondrian.olap.ParameterImpl.new ParameterSlot() {...}
 * Complexity Change:  + 5.26%
 * Instruction Change:  + 8.43%
 * Line Change:  + 17.65%
 * Method Change:  + 9.09%

###### mondrian.olap.Util.Flat2List
 * Branch Change: **-3.70%**:small_red_triangle_down:
 * Complexity Change: **-4.00%**:small_red_triangle_down:
 * Instruction Change: **-.53%**:small_red_triangle_down:

###### mondrian.olap.Util.Flat3List
 * Branch Change: **-2.63%**:small_red_triangle_down:
 * Complexity Change: **-3.23%**:small_red_triangle_down:

###### mondrian.rolap.agg.SegmentCacheManager.AbortException
 * Complexity Change:  + 100.00%
 * Instruction Change:  + 100.00%
 * Line Change:  + 100.00%
 * Method Change:  + 100.00%

###### mondrian.rolap.agg.SegmentCacheManager.Actor
 * Branch Change:  + 25.00%
 * Complexity Change:  + 12.50%
 * Instruction Change:  + 9.80%
 * Line Change:  + 8.62%

###### mondrian.rolap.agg.SegmentLoader
 * Branch Change:  + 2.00%
 * Complexity Change:  + 2.40%
 * Instruction Change:  + 1.29%
 * Line Change:  + .89%

###### mondrian.rolap.agg.SegmentLoader.2.new SegmentCacheManager.Command() {...}
 * Branch Change:  + 33.33%
 * Complexity Change:  + 33.33%
 * Instruction Change:  + 6.78%
 * Line Change:  + 6.67%

###### mondrian.rolap.aggmatcher.AggStar
 * Branch Change: **-1.08%**:small_red_triangle_down:
 * Complexity Change: **-1.30%**:small_red_triangle_down:
 * Instruction Change: **-.74%**:small_red_triangle_down:
 * Line Change: **-1.06%**:small_red_triangle_down:

###### mondrian.rolap.aggmatcher.AggStar.FactTable
 * Branch Change: **-2.26%**:small_red_triangle_down:
 * Complexity Change: **-2.16%**:small_red_triangle_down:
 * Instruction Change: **-2.83%**:small_red_triangle_down:
 * Line Change: **-3.23%**:small_red_triangle_down:
 * Method Change: **-2.63%**:small_red_triangle_down:

###### mondrian.rolap.aggmatcher.DefaultRecognizer
 * Branch Change: **-4.81%**:small_red_triangle_down:
 * Complexity Change: **-3.44%**:small_red_triangle_down:
 * Instruction Change: **-.59%**:small_red_triangle_down:
 * Line Change: **-2.04%**:small_red_triangle_down:

###### mondrian.rolap.aggmatcher.DefaultRecognizer.new Recognizer.Matcher() {...}
 * Complexity New Coverage: 100.00%
 * Instruction New Coverage: 100.00%
 * Line New Coverage: 100.00%
 * Method New Coverage: 100.00%

###### mondrian.rolap.aggmatcher.ExplicitRecognizer
 * Branch Change: **-.89%**:small_red_triangle_down:
 * Complexity Change: **-.54%**:small_red_triangle_down:
 * Instruction Change:  + .70%
 * Line Change:  + .47%

###### mondrian.rolap.aggmatcher.ExplicitRules.TableDef
 * Branch Change: **-1.07%**:small_red_triangle_down:
 * Complexity Change: **-1.13%**:small_red_triangle_down:
 * Instruction Change:  + .06%
 * Line Change:  + .31%
 * Method Change:  + .59%

###### mondrian.rolap.aggmatcher.JdbcSchema.UsageType
 * Instruction Change:  + .58%

###### mondrian.rolap.aggmatcher.Recognizer
 * Branch Change: **-4.54%**:small_red_triangle_down:
 * Complexity Change: **-4.49%**:small_red_triangle_down:
 * Instruction Change: **-3.49%**:small_red_triangle_down:
 * Line Change: **-4.47%**:small_red_triangle_down:
 * Method Change: **-5.93%**:small_red_triangle_down:

###### mondrian.rolap.cache.SegmentCacheIndexImpl
 * Branch Change:  + .47%
 * Complexity Change:  + .73%
 * Instruction Change:  + .05%
 * Line Change:  + .26%

###### mondrian.rolap.CellKey.Three
 * Branch Change: **-5.56%**:small_red_triangle_down:
 * Complexity Change: **-4.55%**:small_red_triangle_down:

###### mondrian.rolap.RolapStatisticsCache
 * Branch Change:  + 7.14%
 * Complexity Change:  + 9.52%
 * Instruction Change:  + .58%
 * Line Change:  + 2.50%

###### mondrian.spi.impl.JdbcStatisticsProvider
 * Branch Change:  + 5.00%
 * Complexity Change:  + 13.33%
 * Instruction Change:  + 6.70%
 * Line Change:  + 4.26%
 * Method Change:  + 20.00%

###### mondrian.util.CacheMap
 * Branch Change:  + 7.14%
 * Complexity Change:  + 4.35%
 * Instruction Change:  + 2.79%
 * Line Change:  + 3.77%

###### mondrian.util.FilteredIterableList
 * Branch Change:  + 5.26%
 * Complexity Change:  + 3.57%
 * Instruction Change:  + .75%

#### License header violations

> :heavy_exclamation_mark: Couldn't find acceptable license on mondrian/src/it/java/mondrian/rolap/aggmatcher/AggMeasureFactCountTest.java

> :heavy_exclamation_mark: Couldn't find acceptable license on mondrian/src/main/java/mondrian/rolap/aggmatcher/AggStar.java

> :heavy_exclamation_mark: Couldn't find acceptable license on mondrian/src/main/java/mondrian/rolap/aggmatcher/DefaultRecognizer.java

> :heavy_exclamation_mark: Couldn't find acceptable license on mondrian/src/main/java/mondrian/rolap/aggmatcher/ExplicitRecognizer.java

> :heavy_exclamation_mark: Couldn't find acceptable license on mondrian/src/main/java/mondrian/rolap/aggmatcher/ExplicitRules.java

> :heavy_exclamation_mark: Couldn't find acceptable license on mondrian/src/main/java/mondrian/rolap/aggmatcher/JdbcSchema.java

> :heavy_exclamation_mark: Couldn't find acceptable license on mondrian/src/main/java/mondrian/rolap/aggmatcher/Recognizer.java

> :heavy_exclamation_mark: Copyright year is not the current year on workbench/src/main/java/mondrian/gui/SchemaExplorer.java. Found 2015. Was expecting 2016

> :heavy_exclamation_mark: Couldn't find acceptable license on workbench/src/main/java/mondrian/gui/SchemaExplorer.java

> :heavy_exclamation_mark: Copyright year is not the current year on workbench/src/main/java/mondrian/gui/SchemaPropertyCellEditor.java. Found 2015. Was expecting 2016

> :heavy_exclamation_mark: Couldn't find acceptable license on workbench/src/main/java/mondrian/gui/SchemaPropertyCellEditor.java

> :heavy_exclamation_mark: Copyright year is not the current year on workbench/src/main/java/mondrian/gui/SchemaTreeModel.java. Found 2015. Was expecting 2016

> :heavy_exclamation_mark: Couldn't find acceptable license on workbench/src/main/java/mondrian/gui/SchemaTreeModel.java

stanislau-strelchanka commented:
1) made fact count required
2) added test cases for fact count (case sensetive, absent etc)
3) fixed tests to check that aggregate tables are used
4) fixed licences, checkstyle.sh doesn't show any error now
5) updated workbench
6) run test suite with aggregates, but skipped Olap4jTckTest.java - it hangs ( there are errors, but they are the same as in master )
           
@kurtwalker , @lucboudreau , @mkambol , @mchen-len-son please review

lucboudreau commented:
About the Olap4jTckTest tests, we need to fix those. It is not ok to have hanging/failing tests, especially not in the API tests.

wingman-pentaho commented:
### Build Completed

:fire: This pull request has some issues. It would be preferable to fix them in order for it to be just perfect. See below for more details. Some links are also available below for further assistance in addressing those issues.

#### Build Commands

```sh
mvn -B -fn -DrunITs=true -f mondrian/pom.xml clean install && mvn -B -f mondrian/pom.xml site
mvn -B -fn -DrunITs=true -f workbench/pom.xml clean install && mvn -B -f workbench/pom.xml site
```

#### Cleanup Commands
```sh
mvn -B -f mondrian/pom.xml build-helper:remove-project-artifact && mvn -B -f workbench/pom.xml build-helper:remove-project-artifact
```

#### Changed files

```console
  mondrian/src/it/java/mondrian/rolap/aggmatcher/AggMeasureFactCountTest.java
  mondrian/src/it/java/mondrian/rolap/aggmatcher/agg_measure_fact_count_test.csv
  mondrian/src/main/java/mondrian/olap/Mondrian.xml
  mondrian/src/main/java/mondrian/rolap/aggmatcher/AggStar.java
  mondrian/src/main/java/mondrian/rolap/aggmatcher/DefaultRecognizer.java
  mondrian/src/main/java/mondrian/rolap/aggmatcher/ExplicitRecognizer.java
  mondrian/src/main/java/mondrian/rolap/aggmatcher/ExplicitRules.java
  mondrian/src/main/java/mondrian/rolap/aggmatcher/JdbcSchema.java
  mondrian/src/main/java/mondrian/rolap/aggmatcher/Recognizer.java
  mondrian/src/site/resources/doc/schema.html
  workbench/src/main/java/mondrian/gui/SchemaExplorer.java
  workbench/src/main/java/mondrian/gui/SchemaPropertyCellEditor.java
  workbench/src/main/java/mondrian/gui/SchemaTreeModel.java
  workbench/src/main/java/mondrian/olap/Mondrian_SW.xml
  workbench/src/main/resources/mondrian/gui/resources/text.properties
```

Still Broken Tests:
======================
###### mondrian.rolap.agg.SegmentBuilderTest.testSameRollupRegardlessOfSegmentOrderNoWildcards:

:red_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: 
SegmentMap is empty. No segmentIds matched test parameters. Full segment cache: [*Segment Header
Schema:[FoodMart]
Checksum:[7b4af973b0d21f364b0a746f5565cb03]
Cube:[Sales]
Measure:[Unit Sales]
Axes:[
    {product_class.product_department=('Alcoholic Beverages','Beverages','Periodicals')}
    {product_class.product_family=('Drink','Non-Consumable')}
    {time_by_day.the_year=('1997')}]
Excluded Regions:[]
Compound Predicates:[]
ID:[63b78b31fae7a3229f284be9a2
```
###### mondrian.rolap.agg.SegmentBuilderTest.testSameRollupRegardlessOfSegmentOrderThreeSegs:

:red_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: 
SegmentMap is empty. No segmentIds matched test parameters. Full segment cache: [*Segment Header
Schema:[FoodMart]
Checksum:[7b4af973b0d21f364b0a746f5565cb03]
Cube:[Sales]
Measure:[Unit Sales]
Axes:[
    {product_class.product_department=('Alcoholic Beverages','Periodicals')}
    {product_class.product_family=('Drink','Non-Consumable')}
    {time_by_day.the_year=('1997')}]
Excluded Regions:[]
Compound Predicates:[]
ID:[e546620837ba618128e2ffbb53f2a32867725f
```
###### mondrian.rolap.agg.SegmentBuilderTest.testSameRollupRegardlessOfSegmentOrderWithData:

:red_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: 
SegmentMap is empty. No segmentIds matched test parameters. Full segment cache: [*Segment Header
Schema:[FoodMart]
Checksum:[7b4af973b0d21f364b0a746f5565cb03]
Cube:[Sales]
Measure:[Unit Sales]
Axes:[
    {product_class.product_department=('Alcoholic Beverages','Baked Goods','Beverages','Periodicals')}
    {product_class.product_family=(*)}
    {time_by_day.the_year=('1997')}]
Excluded Regions:[]
Compound Predicates:[]
ID:[71cf49a1d6f0c2da433442927e05ea5c571
```
###### mondrian.rolap.agg.SegmentBuilderTest.testSameRollupRegardlessOfSegmentOrderWithEmptySegmentBody:

:red_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: 
SegmentMap is empty. No segmentIds matched test parameters. Full segment cache: [*Segment Header
Schema:[FoodMart]
Checksum:[7b4af973b0d21f364b0a746f5565cb03]
Cube:[Sales]
Measure:[Unit Sales]
Axes:[
    {time_by_day.month_of_year=('2','3','4','5','6','7')}
    {time_by_day.quarter=('Q1','Q2','Q3')}
    {time_by_day.the_year=('1998')}]
Excluded Regions:[]
Compound Predicates:[]
ID:[af45d690c892633e5755b06d623ba50a849357832d2157954303a76b4d9564ac]
, *Segment
```

#### Integration test coverage change

These statistics help you identify how your changes have affected the coverage of the following files. If a file is not in this list, then its coverage was not affected by your changes. To get some help interpreting these metrics, please refer to [Jacoco's documentation](http://eclemma.org/jacoco/trunk/doc/counters.html).

###### mondrian.olap.MondrianDef
 * Instruction Change:  + .02%

###### mondrian.olap.MondrianDef.AggName
 * Branch Change: **-1.32%**:small_red_triangle_down:
 * Complexity Change: **-3.24%**:small_red_triangle_down:
 * Instruction Change: **-.96%**:small_red_triangle_down:
 * Line Change:  + .02%

###### mondrian.olap.MondrianDef.AggTable
 * Complexity Change: **-.26%**:small_red_triangle_down:
 * Instruction Change: **-.13%**:small_red_triangle_down:
 * Line Change: **-.16%**:small_red_triangle_down:
 * Method Change:  + 3.21%

###### mondrian.olap.Util.Flat3List
 * Branch Change: **-2.63%**:small_red_triangle_down:
 * Complexity Change: **-3.23%**:small_red_triangle_down:

###### mondrian.rolap.agg.SegmentCacheManager.Actor
 * Instruction Change:  + 4.41%

###### mondrian.rolap.aggmatcher.AggStar
 * Branch Change: **-1.08%**:small_red_triangle_down:
 * Complexity Change: **-1.30%**:small_red_triangle_down:
 * Instruction Change: **-.74%**:small_red_triangle_down:
 * Line Change: **-1.06%**:small_red_triangle_down:

###### mondrian.rolap.aggmatcher.AggStar.FactTable
 * Branch Change: **-2.26%**:small_red_triangle_down:
 * Complexity Change: **-2.16%**:small_red_triangle_down:
 * Instruction Change: **-2.83%**:small_red_triangle_down:
 * Line Change: **-3.23%**:small_red_triangle_down:
 * Method Change: **-2.63%**:small_red_triangle_down:

###### mondrian.rolap.aggmatcher.DefaultRecognizer
 * Branch Change: **-4.81%**:small_red_triangle_down:
 * Complexity Change: **-3.44%**:small_red_triangle_down:
 * Instruction Change: **-.59%**:small_red_triangle_down:
 * Line Change: **-2.04%**:small_red_triangle_down:

###### mondrian.rolap.aggmatcher.DefaultRecognizer.new Recognizer.Matcher() {...}
 * Complexity New Coverage: 100.00%
 * Instruction New Coverage: 100.00%
 * Line New Coverage: 100.00%
 * Method New Coverage: 100.00%

###### mondrian.rolap.aggmatcher.ExplicitRecognizer
 * Branch Change: **-.89%**:small_red_triangle_down:
 * Complexity Change: **-.54%**:small_red_triangle_down:
 * Instruction Change:  + .70%
 * Line Change:  + .47%

###### mondrian.rolap.aggmatcher.ExplicitRules.TableDef
 * Branch Change: **-1.07%**:small_red_triangle_down:
 * Complexity Change: **-1.13%**:small_red_triangle_down:
 * Instruction Change:  + .06%
 * Line Change:  + .31%
 * Method Change:  + .59%

###### mondrian.rolap.aggmatcher.JdbcSchema.UsageType
 * Instruction Change:  + .58%

###### mondrian.rolap.aggmatcher.Recognizer
 * Branch Change: **-4.54%**:small_red_triangle_down:
 * Complexity Change: **-4.49%**:small_red_triangle_down:
 * Instruction Change: **-3.49%**:small_red_triangle_down:
 * Line Change: **-4.47%**:small_red_triangle_down:
 * Method Change: **-5.93%**:small_red_triangle_down:

###### mondrian.rolap.BatchLoader
 * Branch Change:  + 1.00%
 * Complexity Change:  + 1.61%
 * Instruction Change:  + 1.30%
 * Line Change:  + .94%

###### mondrian.rolap.cache.SegmentCacheIndexImpl
 * Branch Change: **-.47%**:small_red_triangle_down:
 * Complexity Change: **-.73%**:small_red_triangle_down:
 * Instruction Change: **-.82%**:small_red_triangle_down:
 * Line Change: **-.26%**:small_red_triangle_down:

###### mondrian.rolap.FastBatchingCellReader
 * Branch Change: **-1.96%**:small_red_triangle_down:
 * Complexity Change: **-3.08%**:small_red_triangle_down:
 * Instruction Change: **-.42%**:small_red_triangle_down:
 * Line Change: **-1.18%**:small_red_triangle_down:

###### mondrian.rolap.RolapStatisticsCache
 * Branch Change:  + 7.14%
 * Complexity Change:  + 9.52%
 * Instruction Change:  + .58%
 * Line Change:  + 2.50%

###### mondrian.server.Execution
 * Instruction Change:  + .76%

###### mondrian.spi.impl.JdbcStatisticsProvider
 * Complexity Change:  + 6.67%
 * Instruction Change:  + 1.03%
 * Line Change:  + 2.13%
 * Method Change:  + 20.00%

###### mondrian.util.CacheMap
 * Branch Change: **-7.14%**:small_red_triangle_down:
 * Complexity Change: **-4.35%**:small_red_triangle_down:
 * Instruction Change: **-2.79%**:small_red_triangle_down:
 * Line Change: **-3.77%**:small_red_triangle_down:

###### mondrian.util.FilteredIterableList
 * Branch Change: **-5.26%**:small_red_triangle_down:
 * Complexity Change: **-3.57%**:small_red_triangle_down:
 * Instruction Change: **-.75%**:small_red_triangle_down:

lucboudreau commented:
Everything checks out. The failures are known and due to the Java 8 change. 

@stanislau-strelchanka Would you mind fixing these 4 tests please? We need to grab the new ID of the corresponding segments and update the Segment test.

wingman-pentaho commented:
### Build Completed

:fire: This pull request has some issues. It would be preferable to fix them in order for it to be just perfect. See below for more details. Some links are also available below for further assistance in addressing those issues.

#### Build Commands

```sh
mvn -B -fn -DrunITs=true -f mondrian/pom.xml clean install && mvn -B -f mondrian/pom.xml site
```

#### Cleanup Commands
```sh
mvn -B -f mondrian/pom.xml build-helper:remove-project-artifact
```

#### Changed files

```console
  mondrian/src/it/java/mondrian/rolap/agg/SegmentBuilderTest.java
```

Still Broken Tests:
======================
###### mondrian.rolap.agg.SegmentBuilderTest.testSameRollupRegardlessOfSegmentOrderNoWildcards:

:red_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: 
SegmentMap is empty. No segmentIds matched test parameters. Full segment cache: [*Segment Header
Schema:[FoodMart]
Checksum:[7b4af973b0d21f364b0a746f5565cb03]
Cube:[Sales]
Measure:[Unit Sales]
Axes:[
    {product_class.product_department=('Alcoholic Beverages','Beverages','Periodicals')}
    {product_class.product_family=('Drink','Non-Consumable')}
    {time_by_day.the_year=('1997')}]
Excluded Regions:[]
Compound Predicates:[]
ID:[63b78b31fae7a3229f284be9a2
```
###### mondrian.rolap.agg.SegmentBuilderTest.testSameRollupRegardlessOfSegmentOrderThreeSegs:

:red_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: 
SegmentMap is empty. No segmentIds matched test parameters. Full segment cache: [*Segment Header
Schema:[FoodMart]
Checksum:[7b4af973b0d21f364b0a746f5565cb03]
Cube:[Sales]
Measure:[Unit Sales]
Axes:[
    {product_class.product_department=('Alcoholic Beverages','Periodicals')}
    {product_class.product_family=('Drink','Non-Consumable')}
    {time_by_day.the_year=('1997')}]
Excluded Regions:[]
Compound Predicates:[]
ID:[e546620837ba618128e2ffbb53f2a32867725f
```
###### mondrian.rolap.agg.SegmentBuilderTest.testSameRollupRegardlessOfSegmentOrderWithData:

:red_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: 
SegmentMap is empty. No segmentIds matched test parameters. Full segment cache: [*Segment Header
Schema:[FoodMart]
Checksum:[7b4af973b0d21f364b0a746f5565cb03]
Cube:[Sales]
Measure:[Unit Sales]
Axes:[
    {product_class.product_department=('Alcoholic Beverages','Baked Goods','Beverages','Periodicals')}
    {product_class.product_family=(*)}
    {time_by_day.the_year=('1997')}]
Excluded Regions:[]
Compound Predicates:[]
ID:[71cf49a1d6f0c2da433442927e05ea5c571
```
###### mondrian.rolap.agg.SegmentBuilderTest.testSameRollupRegardlessOfSegmentOrderWithEmptySegmentBody:

:red_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: 
SegmentMap is empty. No segmentIds matched test parameters. Full segment cache: [*Segment Header
Schema:[FoodMart]
Checksum:[7b4af973b0d21f364b0a746f5565cb03]
Cube:[Sales]
Measure:[Unit Sales]
Axes:[
    {time_by_day.month_of_year=('2','3','4','5','6','7')}
    {time_by_day.quarter=('Q1','Q2','Q3')}
    {time_by_day.the_year=('1998')}]
Excluded Regions:[]
Compound Predicates:[]
ID:[af45d690c892633e5755b06d623ba50a849357832d2157954303a76b4d9564ac]
, *Segment
```

#### Integration test coverage change

These statistics help you identify how your changes have affected the coverage of the following files. If a file is not in this list, then its coverage was not affected by your changes. To get some help interpreting these metrics, please refer to [Jacoco's documentation](http://eclemma.org/jacoco/trunk/doc/counters.html).

###### mondrian.olap.MondrianDef
 * Instruction Change:  + .02%

###### mondrian.olap.MondrianDef.AggName
 * Branch Change: **-1.32%**:small_red_triangle_down:
 * Complexity Change: **-3.24%**:small_red_triangle_down:
 * Instruction Change: **-.96%**:small_red_triangle_down:
 * Line Change:  + .02%

###### mondrian.olap.MondrianDef.AggTable
 * Complexity Change: **-.26%**:small_red_triangle_down:
 * Instruction Change: **-.13%**:small_red_triangle_down:
 * Line Change: **-.16%**:small_red_triangle_down:
 * Method Change:  + 3.21%

###### mondrian.olap.ParameterImpl.new ParameterSlot() {...}
 * Complexity Change: **-5.26%**:small_red_triangle_down:
 * Instruction Change: **-8.43%**:small_red_triangle_down:
 * Line Change: **-17.65%**:small_red_triangle_down:
 * Method Change: **-9.09%**:small_red_triangle_down:

###### mondrian.olap.Util.Flat3List
 * Branch Change: **-2.63%**:small_red_triangle_down:
 * Complexity Change: **-3.23%**:small_red_triangle_down:

###### mondrian.rolap.aggmatcher.AggStar
 * Branch Change: **-1.08%**:small_red_triangle_down:
 * Complexity Change: **-1.30%**:small_red_triangle_down:
 * Instruction Change: **-.74%**:small_red_triangle_down:
 * Line Change: **-1.06%**:small_red_triangle_down:

###### mondrian.rolap.aggmatcher.AggStar.FactTable
 * Branch Change: **-2.26%**:small_red_triangle_down:
 * Complexity Change: **-2.16%**:small_red_triangle_down:
 * Instruction Change: **-2.83%**:small_red_triangle_down:
 * Line Change: **-3.23%**:small_red_triangle_down:
 * Method Change: **-2.63%**:small_red_triangle_down:

###### mondrian.rolap.aggmatcher.DefaultRecognizer
 * Branch Change: **-4.81%**:small_red_triangle_down:
 * Complexity Change: **-3.44%**:small_red_triangle_down:
 * Instruction Change: **-.59%**:small_red_triangle_down:
 * Line Change: **-2.04%**:small_red_triangle_down:

###### mondrian.rolap.aggmatcher.DefaultRecognizer.new Recognizer.Matcher() {...}
 * Complexity New Coverage: 100.00%
 * Instruction New Coverage: 100.00%
 * Line New Coverage: 100.00%
 * Method New Coverage: 100.00%

###### mondrian.rolap.aggmatcher.ExplicitRecognizer
 * Branch Change: **-.89%**:small_red_triangle_down:
 * Complexity Change: **-.54%**:small_red_triangle_down:
 * Instruction Change:  + .70%
 * Line Change:  + .47%

###### mondrian.rolap.aggmatcher.ExplicitRules.TableDef
 * Branch Change: **-1.07%**:small_red_triangle_down:
 * Complexity Change: **-1.13%**:small_red_triangle_down:
 * Instruction Change:  + .06%
 * Line Change:  + .31%
 * Method Change:  + .59%

###### mondrian.rolap.aggmatcher.JdbcSchema.UsageType
 * Instruction Change:  + .58%

###### mondrian.rolap.aggmatcher.Recognizer
 * Branch Change: **-4.54%**:small_red_triangle_down:
 * Complexity Change: **-4.49%**:small_red_triangle_down:
 * Instruction Change: **-3.49%**:small_red_triangle_down:
 * Line Change: **-4.47%**:small_red_triangle_down:
 * Method Change: **-5.93%**:small_red_triangle_down:

###### mondrian.server.Execution
 * Branch Change: **-1.39%**:small_red_triangle_down:
 * Complexity Change: **-1.59%**:small_red_triangle_down:

###### mondrian.util.CacheMap
 * Branch Change: **-7.14%**:small_red_triangle_down:
 * Complexity Change: **-4.35%**:small_red_triangle_down:
 * Instruction Change: **-2.79%**:small_red_triangle_down:
 * Line Change: **-3.77%**:small_red_triangle_down:

###### mondrian.util.FilteredIterableList
 * Branch Change: **-5.26%**:small_red_triangle_down:
 * Complexity Change: **-3.57%**:small_red_triangle_down:
 * Instruction Change: **-.75%**:small_red_triangle_down:

wingman-pentaho commented:
### Build Completed

:x: This pull request has errors. They will need to be addressed before it can be accepted. See below for more details. Some links are also available below for further assistance in addressing those issues.

#### Build Commands

```sh
mvn -B -fn -DrunITs=true -f pom.xml clean install && mvn -B -f pom.xml site
```

#### Cleanup Commands
```sh
mvn -B -f pom.xml build-helper:remove-project-artifact
```

#### Changed files

```console
  mondrian/src/it/java/mondrian/rolap/agg/SegmentBuilderTest.java
```

Newly Broken Tests:
======================
###### mondrian.test.CompoundSlicerTest.testBugMondrian899:

:red_circle: `junit.framework.ComparisonFailure`

```java
junit.framework.ComparisonFailure: 
Expected:
Axis #0:
{[Time].[1997].[Q1].[2]}
{[Time].[1997].[Q1].[3]}
{[Time].[1997].[Q2].[4]}
{[Time].[1997].[Q2].[5]}
{[Time].[1997].[Q2].[6]}
{[Time].[1997].[Q3].[7]}
{[Time].[1997].[Q3].[8]}
{[Time].[1997].[Q3].[9]}
{[Time].[1997].[Q4].[10]}
{[Time].[1997].[Q4].[11]}
Axis #1:
{[Measures].[Unit Sales]}
Axis #2:
{[Customers].[USA].[WA].[Spokane].[Wildon Cameron]}
{[Customers].[USA].[WA].[Spokane].[Emily Barela]}
{[Customers].[USA].[WA].[Spokane].[Dauna Barton
```
###### mondrian.test.CompoundSlicerTest.testCompoundAggCalcMemberInSlicer1:

:red_circle: `junit.framework.ComparisonFailure`

```java
junit.framework.ComparisonFailure: 
Compound aggregated member should return same results with native filter on/off expected:<...Janis Strunk]}
{[Customers].[USA].[OR].[Albany].[L. Albert Wall]}
{[Customers].[USA].[OR].[Albany].[Norman Arnold]}
{[Customers].[USA].[OR].[Albany].[Pauline Kubitschek]}
{[Customers].[USA].[OR].[Albany].[Peggy Denton]}
{[Customers].[USA].[OR].[Albany].[Roslyn Jasper]}
{[Customers].[USA].[OR].[Albany].[Tina Heidi]}
{[Customers].[USA].[OR].[Albany].[Veronica Sulkind]}
{
```
###### mondrian.test.NamedSetTest.testHierarchizeNamedSetImmutable:

:red_circle: `junit.framework.ComparisonFailure`

```java
junit.framework.ComparisonFailure: 
Expected:
Axis #0:
{}
Axis #1:
{[Measures].[Unit Sales]}
Axis #2:
{[Customers].[USA].[WA].[Yakima].[Tracy Meyer], [Store].[USA].[WA].[Yakima].[Store 23]}
{[Customers].[USA].[WA].[Yakima].[Vanessa Thompson], [Store].[USA].[WA].[Yakima].[Store 23]}
{[Customers].[USA].[WA].[Yakima].[Velma Lykes], [Store].[USA].[WA].[Yakima].[Store 23]}
{[Customers].[USA].[WA].[Yakima].[William Battaglia], [Store].[USA].[WA].[Yakima].[Store 23]}
{[Customers].[USA].[WA].[Yakima].[W
```
###### mondrian.test.NamedSetTest.testNamedSetRangeInSlicerPrimed:

:red_circle: `junit.framework.ComparisonFailure`

```java
junit.framework.ComparisonFailure: 
Expected:
Axis #0:
{[Time].[1997].[Q1].[2]}
{[Time].[1997].[Q1].[3]}
{[Time].[1997].[Q2].[4]}
{[Time].[1997].[Q2].[5]}
{[Time].[1997].[Q2].[6]}
{[Time].[1997].[Q3].[7]}
{[Time].[1997].[Q3].[8]}
{[Time].[1997].[Q3].[9]}
{[Time].[1997].[Q4].[10]}
{[Time].[1997].[Q4].[11]}
Axis #1:
{[Measures].[Unit Sales]}
Axis #2:
{[Customers].[USA].[WA].[Spokane].[Wildon Cameron]}
{[Customers].[USA].[WA].[Spokane].[Emily Barela]}
{[Customers].[USA].[WA].[Spokane].[Dauna Barton
```

Newly Fixed Tests:
======================
###### mondrian.rolap.agg.SegmentBuilderTest.testSameRollupRegardlessOfSegmentOrderNoWildcards:

:large_blue_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: 
SegmentMap is empty. No segmentIds matched test parameters. Full segment cache: [*Segment Header
Schema:[FoodMart]
Checksum:[7b4af973b0d21f364b0a746f5565cb03]
Cube:[Sales]
Measure:[Unit Sales]
Axes:[
    {product_class.product_department=('Alcoholic Beverages','Beverages','Periodicals')}
    {product_class.product_family=('Drink','Non-Consumable')}
    {time_by_day.the_year=('1997')}]
Excluded Regions:[]
Compound Predicates:[]
ID:[63b78b31fae7a3229f284be9a2
```
###### mondrian.rolap.agg.SegmentBuilderTest.testSameRollupRegardlessOfSegmentOrderThreeSegs:

:large_blue_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: 
SegmentMap is empty. No segmentIds matched test parameters. Full segment cache: [*Segment Header
Schema:[FoodMart]
Checksum:[7b4af973b0d21f364b0a746f5565cb03]
Cube:[Sales]
Measure:[Unit Sales]
Axes:[
    {product_class.product_department=('Alcoholic Beverages','Periodicals')}
    {product_class.product_family=('Drink','Non-Consumable')}
    {time_by_day.the_year=('1997')}]
Excluded Regions:[]
Compound Predicates:[]
ID:[e546620837ba618128e2ffbb53f2a32867725f
```
###### mondrian.rolap.agg.SegmentBuilderTest.testSameRollupRegardlessOfSegmentOrderWithData:

:large_blue_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: 
SegmentMap is empty. No segmentIds matched test parameters. Full segment cache: [*Segment Header
Schema:[FoodMart]
Checksum:[7b4af973b0d21f364b0a746f5565cb03]
Cube:[Sales]
Measure:[Unit Sales]
Axes:[
    {product_class.product_department=('Alcoholic Beverages','Baked Goods','Beverages','Periodicals')}
    {product_class.product_family=(*)}
    {time_by_day.the_year=('1997')}]
Excluded Regions:[]
Compound Predicates:[]
ID:[71cf49a1d6f0c2da433442927e05ea5c571
```
###### mondrian.rolap.agg.SegmentBuilderTest.testSameRollupRegardlessOfSegmentOrderWithEmptySegmentBody:

:large_blue_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: 
SegmentMap is empty. No segmentIds matched test parameters. Full segment cache: [*Segment Header
Schema:[FoodMart]
Checksum:[7b4af973b0d21f364b0a746f5565cb03]
Cube:[Sales]
Measure:[Unit Sales]
Axes:[
    {time_by_day.month_of_year=('2','3','4','5','6','7')}
    {time_by_day.quarter=('Q1','Q2','Q3')}
    {time_by_day.the_year=('1998')}]
Excluded Regions:[]
Compound Predicates:[]
ID:[af45d690c892633e5755b06d623ba50a849357832d2157954303a76b4d9564ac]
, *Segment
```

#### Integration test coverage change

These statistics help you identify how your changes have affected the coverage of the following files. If a file is not in this list, then its coverage was not affected by your changes. To get some help interpreting these metrics, please refer to [Jacoco's documentation](http://eclemma.org/jacoco/trunk/doc/counters.html).

###### mondrian.olap.MondrianDef
 * Instruction Change:  + .02%

###### mondrian.olap.MondrianDef.AggName
 * Branch Change: **-1.32%**:small_red_triangle_down:
 * Complexity Change: **-3.24%**:small_red_triangle_down:
 * Instruction Change: **-.96%**:small_red_triangle_down:
 * Line Change:  + .02%

###### mondrian.olap.MondrianDef.AggTable
 * Complexity Change: **-.26%**:small_red_triangle_down:
 * Instruction Change: **-.13%**:small_red_triangle_down:
 * Line Change: **-.16%**:small_red_triangle_down:
 * Method Change:  + 3.21%

###### mondrian.olap.ParameterImpl.new ParameterSlot() {...}
 * Complexity Change: **-5.26%**:small_red_triangle_down:
 * Instruction Change: **-8.43%**:small_red_triangle_down:
 * Line Change: **-17.65%**:small_red_triangle_down:
 * Method Change: **-9.09%**:small_red_triangle_down:

###### mondrian.olap.Util
 * Branch Change:  + .16%
 * Complexity Change:  + .21%
 * Instruction Change:  + .02%
 * Line Change:  + .09%

###### mondrian.rolap.agg.SegmentBuilder
 * Branch Change:  + 3.23%
 * Complexity Change:  + 4.30%
 * Instruction Change:  + 1.25%
 * Line Change:  + 1.18%

###### mondrian.rolap.aggmatcher.AggStar
 * Branch Change: **-1.08%**:small_red_triangle_down:
 * Complexity Change: **-1.30%**:small_red_triangle_down:
 * Instruction Change: **-.74%**:small_red_triangle_down:
 * Line Change: **-1.06%**:small_red_triangle_down:

###### mondrian.rolap.aggmatcher.AggStar.FactTable
 * Branch Change: **-2.26%**:small_red_triangle_down:
 * Complexity Change: **-2.16%**:small_red_triangle_down:
 * Instruction Change: **-2.83%**:small_red_triangle_down:
 * Line Change: **-3.23%**:small_red_triangle_down:
 * Method Change: **-2.63%**:small_red_triangle_down:

###### mondrian.rolap.aggmatcher.DefaultRecognizer
 * Branch Change: **-4.81%**:small_red_triangle_down:
 * Complexity Change: **-3.44%**:small_red_triangle_down:
 * Instruction Change: **-.59%**:small_red_triangle_down:
 * Line Change: **-2.04%**:small_red_triangle_down:

###### mondrian.rolap.aggmatcher.DefaultRecognizer.new Recognizer.Matcher() {...}
 * Complexity New Coverage: 100.00%
 * Instruction New Coverage: 100.00%
 * Line New Coverage: 100.00%
 * Method New Coverage: 100.00%

###### mondrian.rolap.aggmatcher.ExplicitRecognizer
 * Branch Change: **-.89%**:small_red_triangle_down:
 * Complexity Change: **-.54%**:small_red_triangle_down:
 * Instruction Change:  + .70%
 * Line Change:  + .47%

###### mondrian.rolap.aggmatcher.ExplicitRules.TableDef
 * Branch Change: **-1.07%**:small_red_triangle_down:
 * Complexity Change: **-1.13%**:small_red_triangle_down:
 * Instruction Change:  + .06%
 * Line Change:  + .31%
 * Method Change:  + .59%

###### mondrian.rolap.aggmatcher.JdbcSchema.UsageType
 * Instruction Change:  + .58%

###### mondrian.rolap.aggmatcher.Recognizer
 * Branch Change: **-4.54%**:small_red_triangle_down:
 * Complexity Change: **-4.49%**:small_red_triangle_down:
 * Instruction Change: **-3.49%**:small_red_triangle_down:
 * Line Change: **-4.47%**:small_red_triangle_down:
 * Method Change: **-5.93%**:small_red_triangle_down:

###### mondrian.rolap.BatchLoader
 * Branch Change:  + 1.00%
 * Complexity Change:  + 1.61%
 * Instruction Change:  + 1.30%
 * Line Change:  + .94%

###### mondrian.rolap.cache.SegmentCacheIndexImpl
 * Branch Change: **-1.40%**:small_red_triangle_down:
 * Complexity Change: **-2.19%**:small_red_triangle_down:
 * Instruction Change: **-1.20%**:small_red_triangle_down:
 * Line Change: **-1.54%**:small_red_triangle_down:

###### mondrian.rolap.FastBatchingCellReader
 * Branch Change: **-1.96%**:small_red_triangle_down:
 * Complexity Change: **-3.08%**:small_red_triangle_down:
 * Instruction Change: **-.42%**:small_red_triangle_down:
 * Line Change: **-1.18%**:small_red_triangle_down:

###### mondrian.rolap.RolapStatisticsCache
 * Branch Change: **-7.14%**:small_red_triangle_down:
 * Complexity Change: **-9.52%**:small_red_triangle_down:
 * Instruction Change: **-.58%**:small_red_triangle_down:
 * Line Change: **-2.50%**:small_red_triangle_down:

###### mondrian.server.MonitorImpl.Handler
 * Branch Change: **-2.27%**:small_red_triangle_down:
 * Complexity Change: **-1.85%**:small_red_triangle_down:
 * Instruction Change: **-.42%**:small_red_triangle_down:
 * Line Change: **-.50%**:small_red_triangle_down:

###### mondrian.spi.impl.JdbcStatisticsProvider
 * Branch Change: **-5.00%**:small_red_triangle_down:
 * Complexity Change: **-13.33%**:small_red_triangle_down:
 * Instruction Change: **-6.70%**:small_red_triangle_down:
 * Line Change: **-4.26%**:small_red_triangle_down:
 * Method Change: **-20.00%**:small_red_triangle_down:

###### mondrian.util.CacheMap
 * Branch Change:  + 7.14%
 * Complexity Change:  + 4.35%
 * Instruction Change:  + 2.79%
 * Line Change:  + 3.77%

###### mondrian.util.FilteredIterableList
 * Branch Change:  + 5.26%
 * Complexity Change:  + 3.57%
 * Instruction Change:  + .75%

stanislau-strelchanka commented:
@kurtwalker , @lucboudreau , @mkambol , @mchen-len-son could you please review. 
I've updated ids in segment test.
Didn't touch implementation code, so not sure what caused newly  broken tests.

wingman-pentaho commented:
### Build Completed

:fire: This pull request has some issues. It would be preferable to fix them in order for it to be just perfect. See below for more details. Some links are also available below for further assistance in addressing those issues.

#### Build Commands

```sh
mvn -B -fn -DrunITs=true -f pom.xml clean install && mvn -B -f pom.xml site
```

#### Cleanup Commands
```sh
mvn -B -f pom.xml build-helper:remove-project-artifact
```

#### Changed files

```console
  mondrian/src/it/java/mondrian/rolap/agg/SegmentBuilderTest.java
```

Newly Fixed Tests:
======================
###### mondrian.rolap.agg.SegmentBuilderTest.testSameRollupRegardlessOfSegmentOrderNoWildcards:

:large_blue_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: 
SegmentMap is empty. No segmentIds matched test parameters. Full segment cache: [*Segment Header
Schema:[FoodMart]
Checksum:[7b4af973b0d21f364b0a746f5565cb03]
Cube:[Sales]
Measure:[Unit Sales]
Axes:[
    {product_class.product_department=('Alcoholic Beverages','Beverages','Periodicals')}
    {product_class.product_family=('Drink','Non-Consumable')}
    {time_by_day.the_year=('1997')}]
Excluded Regions:[]
Compound Predicates:[]
ID:[63b78b31fae7a3229f284be9a2
```
###### mondrian.rolap.agg.SegmentBuilderTest.testSameRollupRegardlessOfSegmentOrderThreeSegs:

:large_blue_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: 
SegmentMap is empty. No segmentIds matched test parameters. Full segment cache: [*Segment Header
Schema:[FoodMart]
Checksum:[7b4af973b0d21f364b0a746f5565cb03]
Cube:[Sales]
Measure:[Unit Sales]
Axes:[
    {product_class.product_department=('Alcoholic Beverages','Periodicals')}
    {product_class.product_family=('Drink','Non-Consumable')}
    {time_by_day.the_year=('1997')}]
Excluded Regions:[]
Compound Predicates:[]
ID:[e546620837ba618128e2ffbb53f2a32867725f
```
###### mondrian.rolap.agg.SegmentBuilderTest.testSameRollupRegardlessOfSegmentOrderWithData:

:large_blue_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: 
SegmentMap is empty. No segmentIds matched test parameters. Full segment cache: [*Segment Header
Schema:[FoodMart]
Checksum:[7b4af973b0d21f364b0a746f5565cb03]
Cube:[Sales]
Measure:[Unit Sales]
Axes:[
    {product_class.product_department=('Alcoholic Beverages','Baked Goods','Beverages','Periodicals')}
    {product_class.product_family=(*)}
    {time_by_day.the_year=('1997')}]
Excluded Regions:[]
Compound Predicates:[]
ID:[71cf49a1d6f0c2da433442927e05ea5c571
```
###### mondrian.rolap.agg.SegmentBuilderTest.testSameRollupRegardlessOfSegmentOrderWithEmptySegmentBody:

:large_blue_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: 
SegmentMap is empty. No segmentIds matched test parameters. Full segment cache: [*Segment Header
Schema:[FoodMart]
Checksum:[7b4af973b0d21f364b0a746f5565cb03]
Cube:[Sales]
Measure:[Unit Sales]
Axes:[
    {time_by_day.month_of_year=('2','3','4','5','6','7')}
    {time_by_day.quarter=('Q1','Q2','Q3')}
    {time_by_day.the_year=('1998')}]
Excluded Regions:[]
Compound Predicates:[]
ID:[af45d690c892633e5755b06d623ba50a849357832d2157954303a76b4d9564ac]
, *Segment
```

#### Integration test coverage change

These statistics help you identify how your changes have affected the coverage of the following files. If a file is not in this list, then its coverage was not affected by your changes. To get some help interpreting these metrics, please refer to [Jacoco's documentation](http://eclemma.org/jacoco/trunk/doc/counters.html).

###### mondrian.olap.MondrianDef
 * Instruction Change:  + .02%

###### mondrian.olap.MondrianDef.AggName
 * Branch Change: **-1.32%**:small_red_triangle_down:
 * Complexity Change: **-3.24%**:small_red_triangle_down:
 * Instruction Change: **-.96%**:small_red_triangle_down:
 * Line Change:  + .02%

###### mondrian.olap.MondrianDef.AggTable
 * Complexity Change: **-.26%**:small_red_triangle_down:
 * Instruction Change: **-.13%**:small_red_triangle_down:
 * Line Change: **-.16%**:small_red_triangle_down:
 * Method Change:  + 3.21%

###### mondrian.olap.Util
 * Branch Change:  + .16%
 * Complexity Change:  + .21%
 * Instruction Change:  + .02%
 * Line Change:  + .09%

###### mondrian.olap.Util.Flat3List
 * Branch Change:  + 2.63%
 * Complexity Change:  + 3.23%

###### mondrian.rolap.agg.AggregationKey
 * Branch Change:  + 3.57%
 * Complexity Change:  + 4.17%
 * Instruction Change:  + 1.15%
 * Line Change:  + 2.38%

###### mondrian.rolap.agg.SegmentBuilder
 * Branch Change:  + 3.23%
 * Complexity Change:  + 4.30%
 * Instruction Change:  + 1.25%
 * Line Change:  + 1.18%

###### mondrian.rolap.agg.SegmentCacheManager.Actor
 * Instruction Change: **-1.47%**:small_red_triangle_down:
 * Line Change: **-3.45%**:small_red_triangle_down:

###### mondrian.rolap.aggmatcher.AggStar
 * Branch Change: **-1.08%**:small_red_triangle_down:
 * Complexity Change: **-1.30%**:small_red_triangle_down:
 * Instruction Change: **-.74%**:small_red_triangle_down:
 * Line Change: **-1.06%**:small_red_triangle_down:

###### mondrian.rolap.aggmatcher.AggStar.FactTable
 * Branch Change: **-2.26%**:small_red_triangle_down:
 * Complexity Change: **-2.16%**:small_red_triangle_down:
 * Instruction Change: **-2.83%**:small_red_triangle_down:
 * Line Change: **-3.23%**:small_red_triangle_down:
 * Method Change: **-2.63%**:small_red_triangle_down:

###### mondrian.rolap.aggmatcher.DefaultRecognizer
 * Branch Change: **-4.81%**:small_red_triangle_down:
 * Complexity Change: **-3.44%**:small_red_triangle_down:
 * Instruction Change: **-.59%**:small_red_triangle_down:
 * Line Change: **-2.04%**:small_red_triangle_down:

###### mondrian.rolap.aggmatcher.DefaultRecognizer.new Recognizer.Matcher() {...}
 * Complexity New Coverage: 100.00%
 * Instruction New Coverage: 100.00%
 * Line New Coverage: 100.00%
 * Method New Coverage: 100.00%

###### mondrian.rolap.aggmatcher.ExplicitRecognizer
 * Branch Change: **-.89%**:small_red_triangle_down:
 * Complexity Change: **-.54%**:small_red_triangle_down:
 * Instruction Change:  + .70%
 * Line Change:  + .47%

###### mondrian.rolap.aggmatcher.ExplicitRules.TableDef
 * Branch Change: **-1.07%**:small_red_triangle_down:
 * Complexity Change: **-1.13%**:small_red_triangle_down:
 * Instruction Change:  + .06%
 * Line Change:  + .31%
 * Method Change:  + .59%

###### mondrian.rolap.aggmatcher.JdbcSchema.UsageType
 * Instruction Change:  + .58%

###### mondrian.rolap.aggmatcher.Recognizer
 * Branch Change: **-4.54%**:small_red_triangle_down:
 * Complexity Change: **-4.49%**:small_red_triangle_down:
 * Instruction Change: **-3.49%**:small_red_triangle_down:
 * Line Change: **-4.47%**:small_red_triangle_down:
 * Method Change: **-5.93%**:small_red_triangle_down:

###### mondrian.rolap.BatchLoader
 * Branch Change:  + 1.00%
 * Complexity Change:  + 1.61%
 * Instruction Change:  + 1.30%
 * Line Change:  + .94%

###### mondrian.rolap.cache.SegmentCacheIndexImpl
 * Branch Change: **-1.87%**:small_red_triangle_down:
 * Complexity Change: **-2.92%**:small_red_triangle_down:
 * Instruction Change: **-2.01%**:small_red_triangle_down:
 * Line Change: **-1.80%**:small_red_triangle_down:

###### mondrian.rolap.CellKey.Three
 * Branch Change:  + 5.56%
 * Complexity Change:  + 4.55%

###### mondrian.rolap.FastBatchingCellReader
 * Branch Change: **-1.96%**:small_red_triangle_down:
 * Complexity Change: **-3.08%**:small_red_triangle_down:
 * Instruction Change: **-.42%**:small_red_triangle_down:
 * Line Change: **-1.18%**:small_red_triangle_down:

###### mondrian.server.Execution
 * Branch Change: **-1.39%**:small_red_triangle_down:
 * Complexity Change: **-1.59%**:small_red_triangle_down:

stanislau-strelchanka commented:
@kurtwalker , @lucboudreau , @mkambol , @mchen-len-son please review

lucboudreau commented:
```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Pentaho Community Edition Project: mondrian ........ SUCCESS [17:52 min]
[INFO] Pentaho Community Edition Project: workbench ....... SUCCESS [ 35.706 s]
[INFO] Pentaho Mondrian Parent POM ........................ SUCCESS [  0.016 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 18:31 min
```
